crates/resolver/src/h3.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2015-2022 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::future::Future;
+use std::net::SocketAddr;
+
+use crate::config::TlsClientConfig;
+use crate::tls::CLIENT_CONFIG;
+
+use proto::h3::{H3ClientConnect, H3ClientStream};
+use proto::xfer::{DnsExchange, DnsExchangeConnect};
+use proto::TokioTime;
+
+use rustls::ClientConfig as CryptoConfig;
+use trust_dns_proto::udp::{DnsUdpSocket, QuicLocalAddr};
+
+#[allow(clippy::type_complexity)]
+#[allow(unused)]
+pub(crate) fn new_h3_stream(
+    socket_addr: SocketAddr,
+    bind_addr: Option<SocketAddr>,
+    dns_name: String,
+    client_config: Option<TlsClientConfig>,
+) -> DnsExchangeConnect<H3ClientConnect, H3ClientStream, TokioTime> {
+    let client_config = if let Some(TlsClientConfig(client_config)) = client_config {
+        client_config
+    } else {
+        match CLIENT_CONFIG.clone() {
+            Ok(client_config) => client_config,
+            Err(error) => return DnsExchange::error(error),
+        }
+    };
+
+    let mut h3_builder = H3ClientStream::builder();
+
+    // TODO: normalize the crypto config settings, can we just use common ALPN settings?
```

Comment:
So we set the ALPN here only if it is unset currently in this PR: https://github.com/bluejekyll/trust-dns/pull/1987/files#diff-ab330120847bfd5337d082f8c41a246b6ddd896dd1ad8a1bb1b7409a271fffdfR379 

It's probably at least worth mentioning that in the docs on the `new_h3_stream`

Reply:
This is the same behavior as in DoQ and DoH. Let me know if you want me to still add this and if I should add it in DoQ and DoH as well.

Reply:
Right, I'm not saying the behavior is different, I'm just commenting on the fact that we don't have the behavior of how the ALPN is set documented and if someone passes in their own ClientConfig with a ALPN specified they might get surprising and hard to detect behavior. I kinda want to suggest here that we recommend not setting ALPN as that will be handled by the underlying protocol, and to only set it if you know that you need to. Something to that extent (and yes, it would be great if we documented that everywhere that it matters).

Reply:
https://github.com/bluejekyll/trust-dns/blob/87a9b6eeabf926ffee18ea97f1246452b111eb29/crates/resolver/src/config.rs#L478-L481

Let me know what you think.

Reply:
Looks good to me.